dev/scripts/generate-tarball.sh

```diff
@@ -13,50 +13,83 @@
-  mvn -T 4C clean install -Dmaven.javadoc.skip=true -DskipTests -Dlicense.skip=true -Dcheckstyle.skip=true -Dfindbugs.skip=true -Pmesos -P${PROFILE} ${BUILD_OPTS} | tee ${BUILD_LOG} 2>&1
-  VERSION=$(bin/alluxio version)
-  mkdir -p ${CLIENT_DIR}/${PROFILE}
-  cp core/client/runtime/target/alluxio-core-client-runtime-${VERSION}-jar-with-dependencies.jar ${CLIENT_DIR}/${PROFILE}/alluxio-${VERSION}-${PROFILE}-client.jar
-done
-PREFIX=alluxio-${VERSION}
-
-cd ..
-rm -rf ${PREFIX}
-mv alluxio ${PREFIX}
-TARGET=${TARBALL_DIR}/${PREFIX}.tar.gz
-
-gtar -czf ${TARGET} ${PREFIX} --exclude-vcs
-popd > /dev/null
-
-echo "Generated tarball at ${TARGET}"
+readonly SCRIPT_DIR=$(cd "$( dirname "$0" )"; pwd)
+readonly TARBALL_DIR="${SCRIPT_DIR}/tarballs"
+readonly WORK_DIR="${SCRIPT_DIR}/workdir"
+readonly CLIENT_DIR="client"
+readonly FRAMEWORKS=( "flink" "presto" "spark" "hadoop" )
+readonly HOME="${SCRIPT_DIR}/../.."
+readonly BUILD_LOG="${HOME}/logs/build.log"
+readonly REPO_COPY="${WORK_DIR}/alluxio"
+
+# Cleans out previous builds and creates a clean copy of the repo.
+function prepare_repo {
+  cd "${SCRIPT_DIR}"
+  mkdir -p "${TARBALL_DIR}"
+  mkdir -p "${WORK_DIR}"
+  rm -rf "${REPO_COPY}"
+  rsync -aq --exclude='logs' --exclude='dev' "${HOME}" "${REPO_COPY}"
+  cd "${REPO_COPY}"
+  git clean -qfdX
+}
+
+function build_framework_clients {
+  cd "${REPO_COPY}" > /dev/null
+  mkdir -p "${CLIENT_DIR}"
+  for profile in "${FRAMEWORKS[@]}"; do
+    echo "Running build ${profile} and logging to ${BUILD_LOG}"
+    mvn -T 4C clean install -Dmaven.javadoc.skip=true -DskipTests -Dlicense.skip=true -Dcheckstyle.skip=true -Dfindbugs.skip=true -Pmesos -P${profile} ${BUILD_OPTS} > "${BUILD_LOG}" 2>&1
+    mkdir -p "${CLIENT_DIR}/${profile}"
+    local version=$(bin/alluxio version)
+    cp "core/client/runtime/target/alluxio-core-client-runtime-${version}-jar-with-dependencies.jar" "${CLIENT_DIR}/${profile}/alluxio-${version}-${profile}-client.jar"
+  done
+}
+
+function build_default {
+  cd "${REPO_COPY}"
+  echo "Running default build and logging to ${BUILD_LOG}"
+  mvn -T 4C clean install -Dmaven.javadoc.skip=true -DskipTests -Dlicense.skip=true -Dcheckstyle.skip=true -Dfindbugs.skip=true -Pmesos ${BUILD_OPTS} > "${BUILD_LOG}" 2>&1
+}
+
+function create_tarball {
+  cd "${REPO_COPY}"
+  local version="$(bin/alluxio version)"
+  local prefix="alluxio-${version}"
+  cd ..
+  rm -rf "${prefix}"
+  mv "${REPO_COPY}" "${prefix}"
+  local target="${TARBALL_DIR}/${prefix}.tar.gz"
+  gtar -czf "${target}" "${prefix}" --exclude-vcs
+  echo "Generated tarball at ${target}"
+}
+
+function main {
+  local build_frameworks=true
+  while [[ "$#" > 0 ]]; do
+    case $1 in
+      --skipFrameworks) build_frameworks=false; shift ;;
+      *) echo "Unrecognized option: $1"; exit 1 ;;
+    esac
+  done
+  prepare_repo
+  if [[ "${build_frameworks}" == true ]]; then
+    build_framework_clients
+  fi
+  build_default
+  create_tarball
+}
```

Comment:
Could we add an option to generate tarballs for all the existing profiles (`hadoop-1` ... `hadoop-2.8`)?